File: native/WinMix.Audio/src/AudioSessionService.cpp
#include "winmix/audio/AudioSessionService.h"
#include "winmix/audio/SessionNaming.h"

#include <endpointvolume.h>
#include <propsys.h>
#include <propidl.h>
#include <functiondiscoverykeys_devpkey.h>

#include <algorithm>
#include <functional>
#include <stdexcept>
#include <unordered_set>

using Microsoft::WRL::ComPtr;

namespace winmix::audio {

namespace {

// Native COM failures surface uniformly as a failed HRESULT -- there is no
// native equivalent of .NET's separate COMException/ObjectDisposedException/
// InvalidCastException distinction, so Guarded() below swallows any
// ComException the same way the original code swallowed all three.
struct ComException : std::runtime_error
{
    HRESULT hr;
    ComException(HRESULT hr_, const char* what) : std::runtime_error(what), hr(hr_) {}
};

void ThrowIfFailed(HRESULT hr, const char* what)
{
    if (FAILED(hr))
    {
        throw ComException(hr, what);
    }
}

// Runs a Core Audio mutation, swallowing the failure that happens when the
// target process exits between a refresh and the user moving its slider.
void Guarded(const std::function<void()>& action)
{
    try
    {
        action();
    }
    catch (const ComException&)
    {
        // The session died underneath us; the next refresh drops the row.
    }
}

std::optional<std::wstring> TakeCoString(LPWSTR raw)
{
    if (raw == nullptr)
    {
        return std::nullopt;
    }
    std::wstring result(raw);
    CoTaskMemFree(raw);
    return result;
}

// Culture-aware, case-insensitive comparison, matching
// StringComparer.CurrentCultureIgnoreCase closely enough for row ordering.
int CompareDisplayNames(const std::wstring& a, const std::wstring& b)
{
    const int result = CompareStringEx(
        LOCALE_NAME_USER_DEFAULT, LINGUISTIC_IGNORECASE,
        a.c_str(), static_cast<int>(a.size()),
        b.c_str(), static_cast<int>(b.size()),
        nullptr, nullptr, 0);
    return result - CSTR_EQUAL;
}

} // namespace

AudioSessionService::AudioSessionService()
{
    ThrowIfFailed(
        CoCreateInstance(__uuidof(MMDeviceEnumerator), nullptr, CLSCTX_ALL, IID_PPV_ARGS(&enumerator_)),
        "CoCreateInstance(MMDeviceEnumerator)");
}

AudioSessionService::~AudioSessionService() = default;

float AudioSessionService::GetMasterVolume()
{
    ComPtr<IAudioEndpointVolume> endpointVolume;
    ThrowIfFailed(
        Device()->Activate(__uuidof(IAudioEndpointVolume), CLSCTX_ALL, nullptr, &endpointVolume),
        "Activate(IAudioEndpointVolume)");

    float level = 0.0f;
    ThrowIfFailed(endpointVolume->GetMasterVolumeLevelScalar(&level), "GetMasterVolumeLevelScalar");
    return level;
}

void AudioSessionService::SetMasterVolume(float scalar)
{
    scalar = std::clamp(scalar, 0.0f, 1.0f);

    ComPtr<IAudioEndpointVolume> endpointVolume;
    ThrowIfFailed(
        Device()->Activate(__uuidof(IAudioEndpointVolume), CLSCTX_ALL, nullptr, &endpointVolume),
        "Activate(IAudioEndpointVolume)");

    ThrowIfFailed(endpointVolume->SetMasterVolumeLevelScalar(scalar, nullptr), "SetMasterVolumeLevelScalar");
}

bool AudioSessionService::GetMasterMuted()
{
    ComPtr<IAudioEndpointVolume> endpointVolume;
    ThrowIfFailed(
        Device()->Activate(__uuidof(IAudioEndpointVolume), CLSCTX_ALL, nullptr, &endpointVolume),
        "Activate(IAudioEndpointVolume)");

    BOOL muted = FALSE;
    ThrowIfFailed(endpointVolume->GetMute(&muted), "GetMute");
    return muted != FALSE;
}

void AudioSessionService::SetMasterMuted(bool muted)
{
    ComPtr<IAudioEndpointVolume> endpointVolume;
    ThrowIfFailed(
        Device()->Activate(__uuidof(IAudioEndpointVolume), CLSCTX_ALL, nullptr, &endpointVolume),
        "Activate(IAudioEndpointVolume)");

    ThrowIfFailed(endpointVolume->SetMute(muted ? TRUE : FALSE, nullptr), "SetMute");
}

std::vector<AudioSessionSnapshot> AudioSessionService::Refresh()
{
    IMMDevice* device = Device();

    ComPtr<IAudioSessionManager2> manager;
    ThrowIfFailed(
        device->Activate(__uuidof(IAudioSessionManager2), CLSCTX_ALL, nullptr, &manager),
        "Activate(IAudioSessionManager2)");

    // Unlike NAudio's AudioSessionManager.RefreshSessions() (a managed-side
    // convenience with no native counterpart), IAudioSessionManager2 has no
    // RefreshSessions method -- GetSessionEnumerator() itself always returns
    // the current session set.
    ComPtr<IAudioSessionEnumerator> sessionEnum;
    ThrowIfFailed(manager->GetSessionEnumerator(&sessionEnum), "GetSessionEnumerator");

    // The previous batch of controls is released here (ComPtr's destructor
    // does the Release) before the map is repopulated below.
    ReleaseControls();

    int count = 0;
    ThrowIfFailed(sessionEnum->GetCount(&count), "GetCount");

    std::vector<AudioSessionSnapshot> snapshots;
    snapshots.reserve(static_cast<size_t>(count));
    std::unordered_set<uint32_t> livePids;

    for (int i = 0; i < count; ++i)
    {
        ComPtr<IAudioSessionControl> baseControl;
        if (FAILED(sessionEnum->GetSession(i, &baseControl)))
        {
            continue;
        }

        ComPtr<IAudioSessionControl2> control;
        if (FAILED(baseControl.As(&control)))
        {
            continue;
        }

        ::AudioSessionState state = ::AudioSessionStateInactive;
        if (FAILED(control->GetState(&state)))
        {
            continue;
        }

        if (state == ::AudioSessionStateExpired)
        {
            continue;
        }

        LPWSTR rawInstanceId = nullptr;
        if (FAILED(control->GetSessionInstanceIdentifier(&rawInstanceId)))
        {
            continue;
        }
        const auto instanceIdOpt = TakeCoString(rawInstanceId);
        if (!instanceIdOpt || instanceIdOpt->empty() || controls_.contains(*instanceIdOpt))
        {
            continue;
        }
        const std::wstring instanceId = *instanceIdOpt;

        DWORD pid = 0;
        control->GetProcessId(&pid);
        livePids.insert(pid);

        const ProcessInfo& info = names_.Get(pid);
        const bool isSystemSounds = control->IsSystemSoundsSession() == S_OK;

        float volume = 0.0f;
        BOOL muted = FALSE;
        ComPtr<ISimpleAudioVolume> simpleVolume;
        if (SUCCEEDED(control->QueryInterface(IID_PPV_ARGS(&simpleVolume))))
        {
            simpleVolume->GetMasterVolume(&volume);
            simpleVolume->GetMute(&muted);
        }

        AudioSessionSnapshot snapshot;
        snapshot.instanceId = instanceId;
        snapshot.pid = pid;
        snapshot.displayName = ChooseDisplayName(control.Get(), info, isSystemSounds);
        snapshot.executablePath = info.executablePath;
        snapshot.volume = volume;
        snapshot.isMuted = muted != FALSE;
        snapshot.peakLevel = ReadPeak(control.Get());
        snapshot.isSystemSounds = isSystemSounds;
        snapshot.state = static_cast<SessionState>(state);

        controls_.emplace(instanceId, control);
        snapshots.push_back(std::move(snapshot));
    }

    names_.Trim(livePids);

    std::sort(snapshots.begin(), snapshots.end(), [](const AudioSessionSnapshot& a, const AudioSessionSnapshot& b)
    {
        if (a.IsActive() != b.IsActive())
        {
            return a.IsActive() && !b.IsActive();
        }
        return CompareDisplayNames(a.displayName, b.displayName) < 0;
    });

    return snapshots;
}

void AudioSessionService::SetVolume(const std::wstring& instanceId, float scalar)
{
    ComPtr<IAudioSessionControl2> control;
    if (!TryGetControl(instanceId, control))
    {
        return;
    }

    scalar = std::clamp(scalar, 0.0f, 1.0f);
    Guarded([&]()
    {
        ComPtr<ISimpleAudioVolume> simpleVolume;
        ThrowIfFailed(control->QueryInterface(IID_PPV_ARGS(&simpleVolume)), "QI(ISimpleAudioVolume)");
        ThrowIfFailed(simpleVolume->SetMasterVolume(scalar, nullptr), "SetMasterVolume");
    });
}

void AudioSessionService::SetMute(const std::wstring& instanceId, bool muted)
{
    ComPtr<IAudioSessionControl2> control;
    if (!TryGetControl(instanceId, control))
    {
        return;
    }

    Guarded([&]()
    {
        ComPtr<ISimpleAudioVolume> simpleVolume;
        ThrowIfFailed(control->QueryInterface(IID_PPV_ARGS(&simpleVolume)), "QI(ISimpleAudioVolume)");
        ThrowIfFailed(simpleVolume->SetMute(muted ? TRUE : FALSE, nullptr), "SetMute");
    });
}

std::vector<AudioDeviceInfo> AudioSessionService::ListOutputDevices()
{
    IMMDevice* device = Device();

    LPWSTR rawDefaultId = nullptr;
    ThrowIfFailed(device->GetId(&rawDefaultId), "GetId");
    const std::wstring defaultId = *TakeCoString(rawDefaultId);

    ComPtr<IMMDeviceCollection> collection;
    ThrowIfFailed(enumerator_->EnumAudioEndpoints(eRender, DEVICE_STATE_ACTIVE, &collection), "EnumAudioEndpoints(render)");

    UINT count = 0;
    ThrowIfFailed(collection->GetCount(&count), "GetCount");

    std::vector<AudioDeviceInfo> devices;
    devices.reserve(count);
    for (UINT i = 0; i < count; ++i)
    {
        ComPtr<IMMDevice> item;
        if (FAILED(collection->Item(i, &item)))
        {
            continue;
        }

        LPWSTR rawId = nullptr;
        if (FAILED(item->GetId(&rawId)))
        {
            continue;
        }
        const std::wstring id = *TakeCoString(rawId);

        devices.push_back(AudioDeviceInfo{id, GetCachedFriendlyName(item.Get(), id), id == defaultId});
    }

    return devices;
}

std::vector<AudioDeviceInfo> AudioSessionService::ListInputDevices()
{
    std::optional<std::wstring> defaultId;
    ComPtr<IMMDevice> defaultDevice;
    if (SUCCEEDED(enumerator_->GetDefaultAudioEndpoint(eCapture, eMultimedia, &defaultDevice)))
    {
        LPWSTR raw = nullptr;
        if (SUCCEEDED(defaultDevice->GetId(&raw)))
        {
            defaultId = TakeCoString(raw);
        }
    }
    // No active capture device at all is not an error here; every row below
    // simply reports isDefault = false.

    ComPtr<IMMDeviceCollection> collection;
    ThrowIfFailed(enumerator_->EnumAudioEndpoints(eCapture, DEVICE_STATE_ACTIVE, &collection), "EnumAudioEndpoints(capture)");

    UINT count = 0;
    ThrowIfFailed(collection->GetCount(&count), "GetCount");

    std::vector<AudioDeviceInfo> devices;
    devices.reserve(count);
    for (UINT i = 0; i < count; ++i)
    {
        ComPtr<IMMDevice> item;
        if (FAILED(collection->Item(i, &item)))
        {
            continue;
        }

        LPWSTR rawId = nullptr;
        if (FAILED(item->GetId(&rawId)))
        {
            continue;
        }
        const std::wstring id = *TakeCoString(rawId);

        devices.push_back(AudioDeviceInfo{id, GetCachedFriendlyName(item.Get(), id), defaultId.has_value() && id == *defaultId});
    }

    return devices;
}

std::wstring AudioSessionService::ChooseDisplayName(IAudioSessionControl2* control, const ProcessInfo& info, bool isSystemSounds)
{
    if (isSystemSounds)
    {
        return L"System sounds";
    }

    const auto resolved = SessionNaming::Resolve(TryReadDisplayName(control));
    return resolved ? *resolved : info.friendlyName;
}

std::optional<std::wstring> AudioSessionService::TryReadDisplayName(IAudioSessionControl2* control)
{
    LPWSTR raw = nullptr;
    if (FAILED(control->GetDisplayName(&raw)))
    {
        return std::nullopt;
    }
    return TakeCoString(raw);
}

float AudioSessionService::ReadPeak(IAudioSessionControl2* control)
{
    ComPtr<IAudioMeterInformation> meter;
    if (FAILED(control->QueryInterface(IID_PPV_ARGS(&meter))))
    {
        return 0.0f;
    }

    float peak = 0.0f;
    if (FAILED(meter->GetPeakValue(&peak)))
    {
        return 0.0f;
    }
    return peak;
}

std::wstring AudioSessionService::GetCachedFriendlyName(IMMDevice* device, const std::wstring& id)
{
    auto it = deviceNameCache_.find(id);
    if (it != deviceNameCache_.end())
    {
        return it->second;
    }

    std::wstring name = id;
    ComPtr<IPropertyStore> store;
    if (SUCCEEDED(device->OpenPropertyStore(STGM_READ, &store)))
    {
        PROPVARIANT variant;
        PropVariantInit(&variant);
        if (SUCCEEDED(store->GetValue(PKEY_Device_FriendlyName, &variant)) &&
            variant.vt == VT_LPWSTR && variant.pwszVal != nullptr)
        {
            name = variant.pwszVal;
        }
        PropVariantClear(&variant);
    }

    deviceNameCache_[id] = name;
    return name;
}

bool AudioSessionService::TryGetControl(const std::wstring& instanceId, ComPtr<IAudioSessionControl2>& out)
{
    auto it = controls_.find(instanceId);
    if (it == controls_.end())
    {
        return false;
    }
    out = it->second;
    return true;
}

void AudioSessionService::ReleaseControls()
{
    controls_.clear();
}

IMMDevice* AudioSessionService::Device()
{
    // Re-resolve when the default endpoint changes under us (headphones
    // plugged in, device disabled); the old device pointer keeps returning
    // stale sessions rather than failing loudly.
    if (device_ && IsUsable(device_.Get()))
    {
        return device_.Get();
    }

    ReleaseControls();
    device_.Reset();

    ThrowIfFailed(
        enumerator_->GetDefaultAudioEndpoint(eRender, eMultimedia, &device_),
        "GetDefaultAudioEndpoint(render)");

    return device_.Get();
}

bool AudioSessionService::IsUsable(IMMDevice* device)
{
    DWORD state = 0;
    if (FAILED(device->GetState(&state)))
    {
        return false;
    }
    return state == DEVICE_STATE_ACTIVE;
}

} // namespace winmix::audio

File: src/WinMix.Audio/AudioSessionService.cs
using NAudio.CoreAudioApi;
using NAudio.CoreAudioApi.Interfaces;

namespace WinMix.Audio;

/// <summary>
/// Reads and controls per-application volume on the default render device via WASAPI.
///
/// Threading: every member must be called from a single thread, and for the WPF
/// shell that is the UI thread. This is deliberate for the volume-mixer phase --
/// the Core Audio objects are apartment-bound and the work is a few COM calls per
/// refresh, so a dedicated thread would buy nothing but marshalling bugs. The
/// equalizer phase is different and will need its own real-time thread; see
/// <c>Loopback/</c>.
///
/// Discovery is poll-based rather than event-based. WASAPI does offer
/// <c>IAudioSessionNotification</c>, but its callbacks arrive on an MTA thread and
/// re-entering the session manager from inside one deadlocks. Polling at a couple
/// of hertz sidesteps that entirely and is imperceptible for a mixer.
/// </summary>
public sealed class AudioSessionService : IDisposable
{
    private readonly MMDeviceEnumerator _enumerator = new();
    private readonly ProcessInfoCache _names = new();
    private readonly Dictionary<string, AudioSessionControl> _controls = new();

    // MMDevice.FriendlyName is a fresh IPropertyStore COM round trip every
    // access -- measured at 35-50ms per device on this machine, regardless of
    // device -- while .ID is effectively free. A device's name does not change
    // while it stays plugged in, so caching it by id turns every poll after the
    // first into a handful of cheap calls instead of one slow one per device.
    // No trim needed the way ProcessInfoCache needs one: unlike pids, device
    // ids don't cycle, so this cannot grow unbounded in practice.
    private readonly Dictionary<string, string> _deviceNameCache = new();

    private MMDevice? _device;
    private bool _disposed;

    /// <summary>Device-wide output level as an amplitude scalar in 0..1.</summary>
    public float MasterVolume
    {
        get => Device().AudioEndpointVolume.MasterVolumeLevelScalar;
        set => Device().AudioEndpointVolume.MasterVolumeLevelScalar = Math.Clamp(value, 0f, 1f);
    }

    public bool MasterMuted
    {
        get => Device().AudioEndpointVolume.Mute;
        set => Device().AudioEndpointVolume.Mute = value;
    }

    /// <summary>
    /// Re-enumerates sessions and returns a fresh snapshot of each one worth showing.
    /// Invalidates any <see cref="AudioSessionSnapshot.InstanceId"/> not present in
    /// the result -- callers must not cache ids across refreshes.
    /// </summary>
    public IReadOnlyList<AudioSessionSnapshot> Refresh()
    {
        ObjectDisposedException.ThrowIf(_disposed, this);

        var manager = Device().AudioSessionManager;
        manager.RefreshSessions();
        var sessions = manager.Sessions;

        // NAudio hands out a new wrapper (and a new COM reference) per indexer
        // access, so the previous batch has to be released or we leak steadily.
        ReleaseControls();

        var snapshots = new List<AudioSessionSnapshot>(sessions.Count);
        var livePids = new HashSet<uint>();

        for (var i = 0; i < sessions.Count; i++)
        {
            var control = sessions[i];

            if (control.State == AudioSessionState.AudioSessionStateExpired)
            {
                control.Dispose();
                continue;
            }

            var instanceId = control.GetSessionInstanceIdentifier;
            if (string.IsNullOrEmpty(instanceId) || _controls.ContainsKey(instanceId))
            {
                control.Dispose();
                continue;
            }

            _controls[instanceId] = control;

            var pid = control.GetProcessID;
            livePids.Add(pid);

            var info = _names.Get(pid);
            snapshots.Add(new AudioSessionSnapshot(
                InstanceId: instanceId,
                Pid: pid,
                DisplayName: ChooseDisplayName(control, info),
                ExecutablePath: info.ExecutablePath,
                Volume: control.SimpleAudioVolume.Volume,
                IsMuted: control.SimpleAudioVolume.Mute,
                PeakLevel: ReadPeak(control),
                IsSystemSounds: control.IsSystemSoundsSession,
                State: control.State));
        }

        _names.Trim(livePids);

        return snapshots
            .OrderByDescending(s => s.IsActive)
            .ThenBy(s => s.DisplayName, StringComparer.CurrentCultureIgnoreCase)
            .ToList();
    }

    /// <summary>Sets one session's amplitude scalar. No-op if the session has gone away.</summary>
    public void SetVolume(string instanceId, float scalar)
    {
        if (TryGetControl(instanceId, out var control))
        {
            Guarded(() => control.SimpleAudioVolume.Volume = Math.Clamp(scalar, 0f, 1f));
        }
    }

    /// <summary>Mutes or unmutes one session. No-op if the session has gone away.</summary>
    public void SetMute(string instanceId, bool muted)
    {
        if (TryGetControl(instanceId, out var control))
        {
            Guarded(() => control.SimpleAudioVolume.Mute = muted);
        }
    }

    /// <summary>Active render endpoints, for the output-device picker.</summary>
    public IReadOnlyList<AudioDeviceInfo> ListOutputDevices()
    {
        ObjectDisposedException.ThrowIf(_disposed, this);

        var defaultId = Device().ID;
        var collection = _enumerator.EnumerateAudioEndPoints(DataFlow.Render, DeviceState.Active);

        var devices = new List<AudioDeviceInfo>(collection.Count);
        foreach (var device in collection)
        {
            devices.Add(new AudioDeviceInfo(device.ID, GetCachedFriendlyName(device), device.ID == defaultId));
            device.Dispose();
        }

        return devices;
    }

    /// <summary>
    /// Makes <paramref name="deviceId"/> the system default render device, for
    /// every role, so both this mixer and every other app follow it.
    /// </summary>
    public void SetDefaultOutputDevice(string deviceId)
    {
        ObjectDisposedException.ThrowIf(_disposed, this);
        Guarded(() => DefaultEndpointSwitcher.SetDefault(deviceId));

        // The device we already resolved is still Active, just no longer
        // default -- drop it so the next Device() call re-resolves to the new
        // default instead of quietly continuing to mix the old one.
        ReleaseControls();
        _device?.Dispose();
        _device = null;
    }

    /// <summary>
    /// Pins one app's output to <paramref name="deviceId"/>, or with null clears
    /// the pin so it follows the system default again.
    /// </summary>
    public void SetAppOutputDevice(uint pid, string? deviceId) =>
        Guarded(() => AppOutputRouter.Set(pid, deviceId));

    /// <summary>Active capture endpoints (microphones), for the input-device picker.</summary>
    public IReadOnlyList<AudioDeviceInfo> ListInputDevices()
    {
        ObjectDisposedException.ThrowIf(_disposed, this);

        string? defaultId = null;
        try
        {
            using var defaultDevice = _enumerator.GetDefaultAudioEndpoint(DataFlow.Capture, Role.Multimedia);
            defaultId = defaultDevice.ID;
        }
        catch (System.Runtime.InteropServices.COMException)
        {
            // No active capture device at all; every row below reports IsDefault = false.
        }

        var collection = _enumerator.EnumerateAudioEndPoints(DataFlow.Capture, DeviceState.Active);

        var devices = new List<AudioDeviceInfo>(collection.Count);
        foreach (var device in collection)
        {
            devices.Add(new AudioDeviceInfo(device.ID, GetCachedFriendlyName(device), device.ID == defaultId));
            device.Dispose();
        }

        return devices;
    }

    /// <summary>
    /// Makes <paramref name="deviceId"/> the system default capture device, for
    /// every role -- the same mechanism as <see cref="SetDefaultOutputDevice"/>,
    /// just pointed at a microphone instead of a speaker. Unlike the render side,
    /// no session state depends on which capture device is default, so there is
    /// nothing to re-resolve afterwards.
    /// </summary>
    public void SetDefaultInputDevice(string deviceId) =>
        Guarded(() => DefaultEndpointSwitcher.SetDefault(deviceId));

    /// <summary>
    /// Picks the label for a row, mirroring how the Windows mixer decides: the
    /// session's own name wins when it has one, otherwise the owning executable.
    /// </summary>
    private static string ChooseDisplayName(AudioSessionControl control, ProcessInfo info)
    {
        if (control.IsSystemSoundsSession)
        {
            return "System sounds";
        }

        return SessionNaming.Resolve(TryReadDisplayName(control)) ?? info.FriendlyName;
    }

    private static string? TryReadDisplayName(AudioSessionControl control)
    {
        try
        {
            return control.DisplayName;
        }
        catch (Exception ex) when (IsSessionGone(ex))
        {
            return null;
        }
    }

    private string GetCachedFriendlyName(MMDevice device)
    {
        if (_deviceNameCache.TryGetValue(device.ID, out var cached))
        {
            return cached;
        }

        var name = device.FriendlyName;
        _deviceNameCache[device.ID] = name;
        return name;
    }

    private bool TryGetControl(string instanceId, out AudioSessionControl control)
    {
        ObjectDisposedException.ThrowIf(_disposed, this);
        return _controls.TryGetValue(instanceId, out control!);
    }

    private static float ReadPeak(AudioSessionControl control)
    {
        try
        {
            return control.AudioMeterInformation.MasterPeakValue;
        }
        catch (Exception ex) when (IsSessionGone(ex))
        {
            return 0f;
        }
    }

    /// <summary>
    /// Runs a Core Audio mutation, swallowing the failure that happens when the
    /// target process exits between our refresh and the user moving its slider.
    /// </summary>
    private static void Guarded(Action action)
    {
        try
        {
            action();
        }
        catch (Exception ex) when (IsSessionGone(ex))
        {
            // The session died underneath us; the next refresh drops the row.
        }
    }

    private static bool IsSessionGone(Exception ex) =>
        ex is System.Runtime.InteropServices.COMException
            or ObjectDisposedException
            or InvalidCastException;

    private MMDevice Device()
    {
        ObjectDisposedException.ThrowIf(_disposed, this);

        // Re-resolve when the default endpoint changes under us (headphones
        // plugged in, device disabled); the old MMDevice keeps returning stale
        // sessions rather than failing loudly.
        if (_device is not null && IsUsable(_device))
        {
            return _device;
        }

        ReleaseControls();
        _device?.Dispose();
        _device = _enumerator.GetDefaultAudioEndpoint(DataFlow.Render, Role.Multimedia);
        return _device;
    }

    private static bool IsUsable(MMDevice device)
    {
        try
        {
            return device.State == DeviceState.Active;
        }
        catch (Exception ex) when (IsSessionGone(ex))
        {
            return false;
        }
    }

    private void ReleaseControls()
    {
        foreach (var control in _controls.Values)
        {
            try
            {
                control.Dispose();
            }
            catch (Exception ex) when (IsSessionGone(ex))
            {
                // Already gone.
            }
        }

        _controls.Clear();
    }

    public void Dispose()
    {
        if (_disposed)
        {
            return;
        }

        _disposed = true;
        ReleaseControls();
        _device?.Dispose();
        _enumerator.Dispose();
    }
}
